src/map.cpp

```diff
@@ -470,8 +352,37 @@ void Map::getSpectators(SpectatorVec& spectators, const Position& centerPos, boo
 			maxRangeZ = centerPos.z;
 		}
 
-		getSpectatorsInternal(spectators, centerPos, minRangeX, maxRangeX, minRangeY, maxRangeY, minRangeZ, maxRangeZ,
-		                      onlyPlayers);
+		auto min_y = centerPos.y + minRangeY;
+		auto min_x = centerPos.x + minRangeX;
+		auto max_y = centerPos.y + maxRangeY;
+		auto max_x = centerPos.x + maxRangeX;
+
+		int32_t minoffset = centerPos.getZ() - maxRangeZ;
+		uint16_t x1 = std::min<uint32_t>(0xFFFF, std::max<int32_t>(0, (min_x + minoffset)));
+		uint16_t y1 = std::min<uint32_t>(0xFFFF, std::max<int32_t>(0, (min_y + minoffset)));
+
+		int32_t maxoffset = centerPos.getZ() - minRangeZ;
+		uint16_t x2 = std::min<uint32_t>(0xFFFF, std::max<int32_t>(0, (max_x + maxoffset)));
+		uint16_t y2 = std::min<uint32_t>(0xFFFF, std::max<int32_t>(0, (max_y + maxoffset)));
+
+		for (auto creature : tfs::map::quadtree::find_in_range(x1, y1, x2, y2)) {
+			if (onlyPlayers && !creature->getPlayer()) {
+				continue;
+			}
+
+			const auto& position = creature->getPosition();
+			if (minRangeZ > position.z || maxRangeZ < position.z) {
+				continue;
+			}
+
+			const auto offsetZ = centerPos.getOffsetZ(position);
+			if ((min_y + offsetZ) > position.y || (max_y + offsetZ) < position.y || (min_x + offsetZ) > position.x ||
+			    (max_x + offsetZ) < position.x) {
+				continue;
+			}
+
+			spectators.emplace_back(creature);
+		}
```

Comment:
@ranisalt  can you implement `C++24` in forgotten? This would avoid push twice into the `SpectatorVec` 🥳